### PR TITLE
[Bountry] hlb-CIFAR10 works with tiny torch backend

### PR DESCRIPTION
**Pull request by eitanturok**
**Update:**
Run
```sh
TINY_BACKEND=1 PYTHONPATH="." python examples/hlb_cifar10_torch.py
```
to train hlb-cifar to 93.82% accuracy with the tiny backend. Yay!

Todo:
- [ ] separate PR for Tensor.unfold + testing
- [ ] separate PR for new pytorch ops
- [ ] separate PR for cleanup hlb_cifar10.py
- [ ] multi-gpu support ?
- [ ] add to benchmarks ?

------------

I've added two files:

`examples/hlb_cifar10_torch_original.py`:
I copied and pasted the original hlb_cifar10 code (already written in pytorch). I implemented a couple of pytorch ops for this but had trouble getting it working. No longer focusing on this.

`examples/hlb_cifar10_torch.py`:
I translated `examples/hlb_cifar10.py` from tinygrad to pytorch. The code here is neater and runs successfully with the tiny backend.

### CODE REVIEW

**Comment by eitanturok** [comment]
In `examples/hlb_cifar10_torch_original.py`, `make_net()` now runs!

Commenting out aten.pad made the pad issue go away. (IDK why.)

Implemented
* aten.reflection_pad1d.out
* aten.replication_pad1d.out
* aten.max.dim
* aten.unfold
* aten.equal

I computed torch.cov and torch.eigh in numpy instead of in torch for now...

**Comment by eitanturok** [comment]
`examples/hlb_cifar10_torch.py` computes the same loss as `examples/hlb_cifar10.py`.

Running
`BS=1 RANDOM_CROP=0 RANDOM_FLIP=0 CUTMIX=0 PYTHONPATH="." python examples/hlb_cifar10.py`
results in
```py
 0 4594.68 ms run, 4593.51 ms python,    1.17 ms METAL, 1272.16 loss, 0.000015 LR, 0.19 GB used,      2.27 GFLOPS,     10.43 GOPS
  1  247.02 ms run,  246.00 ms python,    1.02 ms METAL, 1128.03 loss, 0.000030 LR, 0.21 GB used,      6.34 GFLOPS,      1.57 GOPS
  2  107.95 ms run,    3.63 ms python,  104.32 ms METAL, 1185.57 loss, 0.000045 LR, 0.21 GB used,     14.50 GFLOPS,      1.57 GOPS
  3   10.23 ms run,    2.01 ms python,    8.22 ms METAL, 1420.64 loss, 0.000060 LR, 0.21 GB used,    153.00 GFLOPS,      1.57 GOPS
  4   11.22 ms run,    3.13 ms python,    8.09 ms METAL, 1457.86 loss, 0.000075 LR, 0.21 GB used,    139.55 GFLOPS,      1.57 GOPS
  5    8.41 ms run,    1.83 ms python,    6.58 ms METAL, 1699.88 loss, 0.000090 LR, 0.21 GB used,    186.12 GFLOPS,      1.57 GOPS
  6    8.51 ms run,    1.88 ms python,    6.63 ms METAL, 1308.22 loss, 0.000105 LR, 0.21 GB used,    184.05 GFLOPS,      1.57 GOPS
  7    9.07 ms run,    1.69 ms python,    7.38 ms METAL,  516.01 loss, 0.000120 LR, 0.21 GB used,    172.65 GFLOPS,      1.57 GOPS
  8    9.38 ms run,    1.99 ms python,    7.39 ms METAL, 2214.26 loss, 0.000135 LR, 0.21 GB used,    166.98 GFLOPS,      1.57 GOPS
  9    9.23 ms run,    1.87 ms python,    7.36 ms METAL,  564.38 loss, 0.000149 LR, 0.21 GB used,    169.67 GFLOPS,      1.57 GOPS
 10    9.50 ms run,    2.15 ms python,    7.35 ms METAL, 3783.09 loss, 0.000164 LR, 0.21 GB used,    164.74 GFLOPS,      1.57 GOPS
```

And running
`TINY_BACKEND=1 BS=1 RANDOM_CROP=0 RANDOM_FLIP=0 CUTMIX=0 PYTHONPATH="." python examples/hlb_cifar10_torch.py`
results in
```py
 0 3740.43 ms run, 3740.42 ms python,    0.01 ms tiny, 1272.16 loss, 0.000015 LR
  1  390.99 ms run,  390.99 ms python,    0.00 ms tiny, 1128.03 loss, 0.000030 LR
  2  256.40 ms run,  256.40 ms python,    0.00 ms tiny, 1172.82 loss, 0.000045 LR
  3  338.61 ms run,  338.61 ms python,    0.00 ms tiny, 1443.51 loss, 0.000060 LR
  4  377.33 ms run,  377.33 ms python,    0.00 ms tiny, 1477.67 loss, 0.000075 LR
  5  269.78 ms run,  269.78 ms python,    0.00 ms tiny, 1688.03 loss, 0.000090 LR
  6  388.84 ms run,  388.84 ms python,    0.00 ms tiny, 1313.62 loss, 0.000105 LR
  7  513.50 ms run,  513.50 ms python,    0.00 ms tiny,  503.73 loss, 0.000120 LR
  8  256.97 ms run,  256.97 ms python,    0.00 ms tiny, 2174.50 loss, 0.000135 LR
  9  252.00 ms run,  252.00 ms python,    0.00 ms tiny,  536.84 loss, 0.000150 LR
 10  283.09 ms run,  283.09 ms python,    0.00 ms tiny, 3699.33 loss, 0.000165 LR
```

**Comment by eitanturok** [comment]
` TINY_BACKEND=1 EVAL_STEPS=100 PYTHONPATH="." RANDOM_CROP=0 RANDOM_FLIP=0 CUTMIX=0 python examples/hlb_cifar10_torch.py` returns
`eval     9350/10240 91.31%,    0.59 val_loss STEP=1000 (in 16071.21 ms)`

**Comment by eitanturok** [comment]
We can now train cifar with the tiny backend! No memory leaks and I caught a tricky device error with cutmix.

The cmd
`TINY_BACKEND=1 PYTHONPATH="." python examples/hlb_cifar10_torch.py`
results in
```py
...
998 1667.73 ms run, 1667.73 ms python,    0.00 ms tiny,  501.85 loss, 0.000086 LR, 2.53 GB used,    393.09 GFLOPS,    655.56 GOPS
999  948.89 ms run,  948.89 ms python,    0.00 ms tiny,  494.17 loss, 0.000082 LR, 2.53 GB used,    690.87 GFLOPS,    655.56 GOPS
shuffling test dataset in 0.00 ms (epoch=0)
eval     9607/10240 93.82%,    0.41 val_loss STEP=1000 (in 14301.42 ms)
```

**Comment by eitanturok** [comment]
cleaner PR here #9430
